File: python/robot-simulator/robot_simulator.py
NORTH = 'N'
EAST  = 'E'
SOUTH = 'S'
WEST  = 'W'

class Robot:

	def __init__(self, direction=NORTH, x=0, y=0):
		self.coordinates = (x, y)
		self.bearing = direction
		self._actions = {
			'A': self.advance,
			'L': self.turn_left,
			'R': self.turn_right
		}
		self._directions = {
			NORTH: lambda (x, y): (x, y+1),
			EAST: lambda (x, y): (x+1, y),
			SOUTH: lambda (x, y): (x, y-1),
			WEST: lambda (x, y): (x-1, y),
		}
		self._bearings = [ NORTH, EAST, SOUTH, WEST ]

	def advance(self):
		self.coordinates = self._directions[self.bearing](self.coordinates)

	def turn_right(self):
		idx = self._bearings.index(self.bearing)
		self.bearing = self._bearings[(idx + 1) % 4]

	def turn_left(self):
		idx = self._bearings.index(self.bearing)
		self.bearing = self._bearings[(idx + 3) % 4]

	def simulate(self, instructions):
		map(lambda i: self._actions[i](), instructions)

File: javascript/robot-simulator/robot-simulator.js
class Robot {

    constructor(bearing = BEARINGS[0]) {
        this.bearing = bearing;
    }

    orient(bearing) {
        if (!BEARINGS.includes(bearing)) {
            throw new Error('Invalid Robot Bearing');
        }
        this.bearing = bearing;
    }

    turnLeft() {
        const idx = BEARINGS.indexOf(this.bearing);
        this.bearing = BEARINGS[(idx + 3) % 4];
    }

    turnRight() {
        const idx = BEARINGS.indexOf(this.bearing);
        this.bearing = BEARINGS[(idx + 1) % 4];
    }

    at(...coords) {
        this.coordinates = coords;
    }

    advance() {
        this.coordinates = DIRECTIONS[this.bearing](...this.coordinates);
    }

    instructions(actions) {
        return [...actions].map(a => ACTIONS[a]);
    }

    place(data) {
        this.at(data.x, data.y);
        this.orient(data.direction);
    }

    evaluate(actions) {
        [...actions].forEach(action => this[ACTIONS[action]]());
    }

}

module.exports = Robot;

const ACTIONS = {
    A: 'advance',
    L: 'turnLeft',
    R: 'turnRight'
};
const DIRECTIONS = {
    north: (x, y) => [x, y + 1],
    east: (x, y) => [x + 1, y],
    south: (x, y) => [x, y - 1],
    west: (x, y) => [x - 1, y]
};
const BEARINGS = Object.keys(DIRECTIONS);